# Day 6: Lanternfish

## Part one

The sea floor is getting steeper. Maybe the sleigh keys got carried this way?

A massive school of glowing *lanternfish* swims past. They must spawn quickly to reach such large numbers - maybe **exponentially** quickly? You should model their growth rate to be sure.

Although you know nothing about this specific species of lanternfish, you make some guesses about their attributes. Surely, each lanternfish creates a new lanternfish once every `7` days.

However, this process isn't necessarily synchronized between every lanternfish - one lanternfish might have 2 days left until it creates another lanternfish, while another might have 4. So, you can model each fish as a single number that represents **the number of days until it creates a new lanternfish**.

Furthermore, you reason, a **new** lanternfish would surely need slightly longer before it's capable of producing more lanternfish: two more days for its first cycle.

So, suppose you have a lanternfish with an internal timer value of 3:

- After one day, its internal timer would become `2`.
- After another day, its internal timer would become `1`.
- After another day, its internal timer would become `0`.
- After another day, its internal timer would reset to `6`, and it would create a **new** lanternfish with an internal timer of `8`.
- After another day, the first lanternfish would have an internal timer of 5, and the second lanternfish would have an internal timer of `7`.

A lanternfish that creates a new fish resets its timer to 6, not 7 (because 0 is included as a valid timer value). The new lanternfish starts with an internal timer of 8 and does not start counting down until the next day.

Realizing what you're trying to do, the submarine automatically produces a list of the ages of several hundred nearby lanternfish (your puzzle input). For example, suppose you were given the following list:

`3,4,3,1,2`

This list means that the first fish has an internal timer of 3, the second fish has an internal timer of 4, and so on until the fifth fish, which has an internal timer of 2. Simulating these fish over several days would proceed as follows:

```
Initial state: 3,4,3,1,2
After  1 day:  2,3,2,0,1
After  2 days: 1,2,1,6,0,8
After  3 days: 0,1,0,5,6,7,8
After  4 days: 6,0,6,4,5,6,7,8,8
After  5 days: 5,6,5,3,4,5,6,7,7,8
After  6 days: 4,5,4,2,3,4,5,6,6,7
After  7 days: 3,4,3,1,2,3,4,5,5,6
After  8 days: 2,3,2,0,1,2,3,4,4,5
After  9 days: 1,2,1,6,0,1,2,3,3,4,8
After 10 days: 0,1,0,5,6,0,1,2,2,3,7,8
After 11 days: 6,0,6,4,5,6,0,1,1,2,6,7,8,8,8
After 12 days: 5,6,5,3,4,5,6,0,0,1,5,6,7,7,7,8,8
After 13 days: 4,5,4,2,3,4,5,6,6,0,4,5,6,6,6,7,7,8,8
After 14 days: 3,4,3,1,2,3,4,5,5,6,3,4,5,5,5,6,6,7,7,8
After 15 days: 2,3,2,0,1,2,3,4,4,5,2,3,4,4,4,5,5,6,6,7
After 16 days: 1,2,1,6,0,1,2,3,3,4,1,2,3,3,3,4,4,5,5,6,8
After 17 days: 0,1,0,5,6,0,1,2,2,3,0,1,2,2,2,3,3,4,4,5,7,8
After 18 days: 6,0,6,4,5,6,0,1,1,2,6,0,1,1,1,2,2,3,3,4,6,7,8,8,8,8
```

Each day, a `0` becomes a `6` and adds a new `8` to the end of the list, while each other number decreases by `1` if it was present at the start of the day.

In this example, after 18 days, there are a total of `26` fish. After 80 days, there would be a total of `5934`.

Find a way to simulate lanternfish. **How many lanternfish would there be after 80 days?**

In [9]:
fishes = []

with open('data/input.txt', 'r') as f:
    line = f.read().splitlines()[0]
    for fish in line.split(','):
        fishes.append(int(fish))
        
fishes[:10]

[3, 5, 3, 1, 4, 4, 5, 5, 2, 1]

In [10]:
def process_one_day(fishes):
    number_of_new_fishes = 0
    new_fishes = []
    for fish in fishes:
        if fish == 0:
            number_of_new_fishes += 1
            new_fishes.append(6)
        else:
            new_fishes.append(fish - 1)
    
    for i in range(0, number_of_new_fishes):
        new_fishes.append(8)
        
    return new_fishes

example_fishes = [3,4,3,1,2]
print('day 0: %s' % (example_fishes, ))

for i in range(1, 8):
    example_fishes = process_one_day(example_fishes)
    print('day %s: %s' % (i, example_fishes, ))

example_fishes = [3,4,3,1,2]

for i in range(1, 81):
    example_fishes = process_one_day(example_fishes)
    
print('After 80 days: %s fishes' % (len(example_fishes), ))

day 0: [3, 4, 3, 1, 2]
day 1: [2, 3, 2, 0, 1]
day 2: [1, 2, 1, 6, 0, 8]
day 3: [0, 1, 0, 5, 6, 7, 8]
day 4: [6, 0, 6, 4, 5, 6, 7, 8, 8]
day 5: [5, 6, 5, 3, 4, 5, 6, 7, 7, 8]
day 6: [4, 5, 4, 2, 3, 4, 5, 6, 6, 7]
day 7: [3, 4, 3, 1, 2, 3, 4, 5, 5, 6]
After 80 days: 5934 fishes


In [11]:
for i in range(0, 80):
    fishes = process_one_day(fishes)
    
print('After 80 days: %s fishes' % (len(fishes), ))

After 80 days: 351092 fishes


Your puzzle answer was `351092`.

## Part two

Suppose the lanternfish live forever and have unlimited food and space. Would they take over the entire ocean?

After 256 days in the example above, there would be a total of `26984457539` lanternfish!

**How many lanternfish would there be after 256 days?**

In [35]:
fishes = [0, 0, 0, 0, 0, 0, 0, 0, 0]

with open('data/input.txt', 'r') as f:
    line = f.read().splitlines()[0]
    for fish in line.split(','):
        fishes[int(fish)] += 1
        
fishes

[0, 83, 51, 58, 54, 54, 0, 0, 0]

In [36]:
def process_one_day_ex2(fishes):
    new_fishes = [
        fishes[1],
        fishes[2],
        fishes[3],
        fishes[4],
        fishes[5],
        fishes[6],
        fishes[7] + fishes[0],
        fishes[8],
        fishes[0],
    ]
    
    return new_fishes

example_fishes = [0, 1, 1, 2, 1, 0, 0, 0, 0]

for i in range(0, 80):
    example_fishes = process_one_day_ex2(example_fishes)
    
print('After 80 days: %s - %s fishes' % (example_fishes, sum(example_fishes), ))

After 80 days: [424, 729, 558, 790, 739, 762, 991, 370, 571] - 5934 fishes


In [37]:
for i in range(0, 80):
    fishes = process_one_day_ex2(fishes)

print('After 80 days: %s - %s fishes' % (fishes, sum(fishes), ))

for i in range(80, 256):
    fishes = process_one_day_ex2(fishes)

print('After 256 days: %s - %s fishes' % (fishes, sum(fishes), ))

After 80 days: [25532, 42594, 31998, 47322, 42341, 47270, 57495, 23297, 33243] - 351092 fishes
After 256 days: [140304164138, 162039539115, 170391060831, 188187916968, 208218433266, 218720012046, 252082120574, 117783213814, 137604155253] - 1595330616005 fishes


Your puzzle answer was `1595330616005`.